shared/engine/session.js

```diff
@@ -0,0 +1,157 @@
+  _endHandler: ?EndHandlerType;
+  // Sequence IDs we've seen. Value is true if we've responded (often we get cancel after we've replied)
+  _seqIDResponded: {[key: string]: boolean} = {};
+  // If you want to know about being cancelled
+  _cancelHandler: ?CancelHandlerType;
+  // If true this session exists forever
+  _dangling: boolean;
+  // Name of the start method, just to help debug
+  _startMethod: ?MethodKey;
+  // Start callback so we can cancel our own callback
+  _startCallback: ?(err: RPCError, ...args: Array<any>) => void;
+
+  // Allow us to make calls
+  _invoke: invokeType;
+
+  // Outstanding requests
+  _outgoingRequests: Array<Object> = [];
+  _incomingRequests: Array<Object> = [];
+
+  constructor (
+    sessionID: SessionID,
+    incomingCallMap: ?incomingCallMapType,
+    waitingHandler: ?WaitingHandlerType,
+    invoke: invokeType,
+    endHandler: EndHandlerType,
+    cancelHandler?: ?CancelHandlerType,
+    dangling?: boolean = false,
+  ) {
+    this._id = sessionID
+    this._incomingCallMap = incomingCallMap || {}
+    this._waitingHandler = waitingHandler
+    this._invoke = invoke
+    this._endHandler = endHandler
+    this._cancelHandler = cancelHandler
+    this._dangling = dangling
+  }
+
+  set id (sessionID: SessionID) { throw new Error("Can't set sessionID") }
+  get id (): SessionID { return this._id }
+  get dangling (): boolean { return this._dangling }
+
+  // Make a waiting handler for the request. We add additional data before calling the parent waitingHandler
+  // and do internal bookkeeping if the request is done
+  _makeWaitingHandler (isOutgoing: boolean, method: MethodKey, seqid: ?number) {
+    return (waiting: boolean) => {
+      rpcLog('engineInternal', 'waiting state change', this.id, waiting, method, this, seqid)
+      // Call the outer handler with all the params it needs
+      this._waitingHandler && this._waitingHandler(waiting, method, this._id)
+
+      // Request is finished, do cleanup
+      if (!waiting) {
+        const requests = isOutgoing ? this._outgoingRequests : this._incomingRequests
+        const idx = requests.findIndex(r => r.method === method)
+        if (idx !== -1) {
+          // Mark us as responded
+          if (seqid) {
+            this._seqIDResponded[String(seqid)] = true
+          }
+          // Remove from our list
+          requests.splice(idx, 1)
+        }
+      }
+    }
+  }
+
+  cancel () {
+    if (this._cancelHandler) {
+      this._cancelHandler(this)
+    } else if (this._startCallback) {
+      this._startCallback({
+        code: ConstantsStatusCode.sccanceled,
+        desc: 'Received RPC cancel for session',
+      })
+    }
+
+    this.end()
+  }
+
+  end () {
+    this._endHandler && this._endHandler(this)
+  }
+
+  // Start the session normally. Tells engine we're done at the end
+  start (method: MethodKey, param: ?Object, callback: ?() => void) {
+    this._startMethod = method
+    this._startCallback = callback
```

Comment:
🐑 🐑 `startCallback` is a bit confusing of a name, but I can't think of a better one